homeassistant/components/supla/__init__.py

```diff
@@ -71,23 +82,54 @@ def setup(hass, base_config):
             )
             return False
 
-    discover_devices(hass, base_config)
+    # Register a cleanup callback
+    async def _async_close_supla_servers(event):
+        for server_name, server in hass.data[DOMAIN][SUPLA_SERVERS].items():
+            _LOGGER.info("Closing up Supla server: %s", server_name)
+            await server.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_close_supla_servers)
+
+    await discover_devices(hass, base_config)
 
     return True
 
 
-def discover_devices(hass, hass_config):
+async def discover_devices(hass, hass_config):
     """
     Run periodically to discover new devices.
 
-    Currently it's only run at startup.
+    Currently it is only run at startup.
     """
     component_configs = {}
 
-    for server_name, server in hass.data[SUPLA_SERVERS].items():
+    for server_name, server in hass.data[DOMAIN][SUPLA_SERVERS].items():
+
+        async def _fetch_channels():
+            async with async_timeout.timeout(30):
```

Comment:
is this a slow process, ideally this wouldnt take longer than scan interval

Reply:
Right, update should be few s max. I;ve changed it to SCAN_INTERVAL (10 s).